lib/Dialect/ESI/Passes/ESILowerBundles.cpp

```diff
@@ -0,0 +1,233 @@
+    converter.createNewOutput(origPort, "_" + ch.name.getValue(), ch.type,
+                              pack ? pack.getFromChannels()[idx] : nullptr,
+                              newOutputChannels[idx]);
+}
+
+/// For an output port, we need to unpack the results from the original value
+/// into the new channel ports.
+void BundlePort::buildOutputSignals() {
+  auto bundleType = cast<ChannelBundleType>(origPort.type);
+  SmallVector<Value, 4> unpackOperands;
+  if (body)
+    unpackOperands.push_back(
+        body->getTerminator()->getOperand(origPort.argNum));
+  SmallVector<BundledChannel, 4> outputChannels;
+
+  SmallVector<Type, 4> unpackOpResultTypes;
+  for (BundledChannel ch : bundleType.getChannels()) {
+    // 'from' on an input bundle becomes an input channel.
+    if (ch.direction == ChannelDirection::from) {
+      hw::PortInfo newPort;
+      unpackOperands.push_back(converter.createNewInput(
+          origPort, "_" + ch.name.getValue(), ch.type, newPort));
+      newInputChannels.push_back(newPort);
+    } else {
+      // 'to' on an input bundle becomes an output channel.
+      unpackOpResultTypes.push_back(ch.type);
+      outputChannels.push_back(ch);
+    }
+  }
+
+  // For an output port, the original bundle must be unpacked into the
+  // individual channel ports.
+  UnpackBundleOp unpack;
+  if (body) {
+    ImplicitLocOpBuilder b(origPort.loc, body, body->end());
+    unpack = b.create<UnpackBundleOp>(unpackOpResultTypes, unpackOperands);
+  }
+
+  // Build new ports and put the new port info directly into the member
+  // variable.
+  newOutputChannels.resize(outputChannels.size());
+  for (auto [idx, ch] : llvm::enumerate(outputChannels))
+    converter.createNewOutput(origPort, "_" + ch.name.getValue(), ch.type,
+                              unpack ? unpack.getToChannels()[idx] : nullptr,
+                              newOutputChannels[idx]);
+}
+
+namespace {
+/// Convert all the ESI bundle ports on modules to channel ports.
+struct ESIBundlesPass : public LowerESIBundlesBase<ESIBundlesPass> {
+  void runOnOperation() override;
+};
+} // anonymous namespace
+
+/// Iterate through the `hw.module[.extern]`s and lower their ports.
+void ESIBundlesPass::runOnOperation() {
+  MLIRContext &ctxt = getContext();
+  ModuleOp top = getOperation();
+
+  // Find all modules and run port conversion on them.
+  circt::hw::InstanceGraph &instanceGraph =
+      getAnalysis<circt::hw::InstanceGraph>();
+  for (auto mod : top.getOps<HWMutableModuleLike>()) {
+    if (failed(PortConverter<ESIBundleConversionBuilder>(instanceGraph, mod)
+                   .run()))
+      return signalPassFailure();
+  }
+
+  // Canonicalize away bundle packs and unpacks. Any non-back-to-back [un]packs
+  // need to be gone by now.
+  ConversionTarget tgt(ctxt);
+  tgt.markUnknownOpDynamicallyLegal([](Operation *) { return true; });
+  tgt.addIllegalOp<PackBundleOp, UnpackBundleOp>();
+  RewritePatternSet patterns(&ctxt);
```

Comment:
I should do this in the Ibis passes that expect a certain canonical form

Reply:
Yep